proposals/p2173.md

```diff
@@ -0,0 +1,1291 @@
+    constraint where the associated constant is ambiguous. Given `T:! X & Y`,
+    the type expression `T.R` is ambiguous between a rewrite to `A` and a
+    rewrite to `B`. But given `T:! X & X`, `T.R` is unambiguously rewritten to
+    `A`.
+-   Combining a constraint with a rewrite rule with a constraint with no rewrite
+    rule preserves the rewrite rule. For example, supposing that
+    `interface Container` extends `interface Iterable`, and `Iterable` has an
+    associated constant `Element`, the constraint
+    `Container & (Iterable where .Element = i32)` is the same as the constraint
+    `(Container & Iterable) where .Element = i32` which is the same as the
+    constraint `Container where .Element = i32`.
+
+### Combining constraints with `and`
+
+It's possible for one `=` constraint in a `where` to refer to another. When this
+happens, the constraint `C where A and B` is interpreted as
+`(C where A) where B`, so rewrites in `A` are applied immediately to names in
+`B`, but rewrites in `B` are not applied to names in `A` until the constraint is
+[resolved](#constraint-resolution):
+
+```
+interface C {
+  let T:! Type;
+  let U:! Type;
+  let V:! Type;
+}
+class M {
+  alias Me = Self;
+}
+// ✅ Same as `C where .T = M and .U = M.Me`, which is
+// the same as `C where .T = M and .U = M`.
+fn F[A:! C where .T = M and .U = .T.Me]() {}
+// ❌ No member `Me` in `A.T:! Type`.
+fn F[A:! C where .U = .T.Me and .T = M]() {}
+```
+
+### Combining constraints with `extends`
+
+Within an interface or named constraint, `extends` can be used to extend a
+constraint that has rewrites.
+
+```
+interface A {
+  let T:! Type;
+  let U:! Type;
+}
+interface B {
+  extends A where .T = .U and .U = i32;
+}
+
+var n: i32;
+
+// ✅ Resolved constraint on `T` is
+// `B where .(A.T) = i32 and .(A.U) = i32`.
+// `T.(A.T)` is rewritten to `i32`.
+fn F(T:! B) -> T.(A.T) { return n; }
+```
+
+### Combining constraints with `impl as`
+
+Within an interface or named constraint, the `impl T as C` syntax permits `=`
+constraints. However, because these constraints don't change the type of `T`,
+rewrite constraints will never result in a rewrite being performed, and instead
+are equivalent to `==` constraints.
+
+```
+interface A {
+  let T:! Type;
+  let U:! Type;
+}
+constraint C {
+  extends A where .T = .U and .U = i32;
+}
+constraint D {
+  extends A where .T == .U and .U == i32;
+}
+interface B {
+  // This constraint is equivalent to
+  // `impl as A where .T == .U and .U == i32;` or
+  // `impl as C;` or `impl as D;`.
+  impl as A where .T = .U and .U = i32;
+}
+
+var n: i32;
+
+// ❌ No implicit conversion from `i32` to `T.(A.T)`.
+// Resolved constraint on `T` is
+// `B where T.(A.T) == T.(A.U) and T.(A.U) = i32`.
+// `T.(A.T)` is single-step equal to `T.(A.U)`, and
+// `T.(A.U)` is single-step equal to `i32`, but
+// `T.(A.T)` is not single-step equal to `i32`.
+fn F(T:! B) -> T.(A.T) { return n; }
+```
```

Comment:
This is perhaps an argument to select the "Rewrite constraints in `impl as` constraints" alternative -- so there isn't a cliff between `extends` and `impl as`. Then we might have this restriction on rewrite constraints only on `impl OtherType as` declarations.

Reply:
> This is perhaps an argument to select the "Rewrite constraints in `impl as` constraints" alternative -- so there isn't a cliff between `extends` and `impl as`. Then we might have this restriction on rewrite constraints only on `impl OtherType as` declarations.

Yeah, that would make sense to me.

Or more generally, preclude rewrite constraints when the type can be something other than Self -- whether it is *spelled* `impl as` or `impl T as` maybe isn't as important as whether we can prove that `T` is exactly `Self`. (If I understood the necessary part correctly?)

My preference to *completely* reject them (rather than allow the syntax but w/o the rewrite semantic) remains though.

Reply:
I think the biggest issue here is rewrite constraints that are found _indirectly_:

```
interface A { let T:! type; }
constraint B { extends C where .T = i32; }
interface C { impl as B; }
```

Here, I think it's probably important that we accept the `impl as B`, even though `B` includes a rewrite constraint that won't end up being part of `C`. Rejecting this case seems likely to be problematic -- it'd make `=` constraints and `impl as` incompatible. As an example, the presence of an `impl as` in a consumer of a constraint would prevent certain forms of library evolution, such as adding a new associated constant with a default value, along with a rewrite constraint for that associated constant in some derived constraint.

We could try to add a _syntactic_ restriction on `where ... = ...` appearing at the top level within an `impl as`. Some choices here, such as rejecting any `=` constraint is written that are written anywhere within an `impl as` declaration, or tracking the provenance of the constraint in some way, seem awkward and somewhat arbitrary. But perhaps we could use a similar rule to the one we picked for scoping / reuse of `.Self`, and only recurse into `where` and `&` expressions and parentheses on the RHS of `impl ? as`, and reject any `=` constraints we find there. I think I'd be OK with that, but we should discuss whether we think that check is better implemented as a language rule or a style rule.

#2476 is also somewhat relevant here.

Reply:
Is that supposed to be `constraint B { extends A where .T = i32; }`?

Reply:
Oops, yes it is :)

Reply:
I'm sufficiently convinced that a syntactic restriction makes sense here that I've added one to the proposal.

Reply:
To summarize my understanding of the discussion so far, given this example:
```
interface A { let T:! type; }
constraint B { extends A where .T = i32; }
interface C { impl as B; }
class D { impl as C {} }
interface E { impl as A where .T = i32; }
interface F(U:! Type) { impl U as B; }
```
We have listed the following options:
- Proposal as-is: the above is accepted with `=` constraints changed to `==` constraints in interfaces `C`, `E`, and `F`. This means `D.(A.T) == i32` but there is no rewrite `D.(A.T)` ~> `i32`.
- We never replace `=` constraints into `==` constraints, but that means rejecting interfaces `C`, `E`, and `F` above.
- We reject `=` constraints in a `impl as __ where ...` syntactically, as in `interface E`, but allow replacing them with `==` constraints when used indirectly as in interfaces `C` and `F`.
- We adopt the "Rewrite constraints in `impl as` constraints" alternative, so interfaces `C` and `E` are accepted and those `=` constraints are consistently rewrite constraints. We then have to select one of the above options for interface `F`.

Reply:
I think that's right, but the proposal as-is is now your third option. I realized the same thing also happens for `where ... is (__ where ...)`, so I've called that case out too; I think the same rules should apply there, because `where T is C` is the `where` equivalent of `impl T as C`.

Reply:
I'm happy with this end state.